mathics/builtin/assignment.py

```diff
@@ -65,376 +74,577 @@ def get_symbol_list(list, error_callback):
+        raise AssignmentException(lhs, None)
+
+
+def process_assign_definition_values(self, lhs, rhs, evaluation, tags, upset):
+    name = lhs.get_head_name()
+    tag = find_tag_and_check(lhs, tags, evaluation)
+    rules = rhs.get_rules_list()
+    if rules is None:
+        evaluation.message(name, "vrule", lhs, rhs)
+        raise AssignmentException(lhs, None)
+    evaluation.definitions.set_values(tag, name, rules)
+    return True
+
+
+def process_assign_options(self, lhs, rhs, evaluation, tags, upset):
+    lhs_leaves = lhs.leaves
+    name = lhs.get_head_name()
+    if len(lhs_leaves) != 1:
+        evaluation.message_args(name, len(lhs_leaves), 1)
+        raise AssignmentException(lhs, rhs)
+    tag = lhs_leaves[0].get_name()
+    if not tag:
+        evaluation.message(name, "sym", lhs_leaves[0], 1)
+        raise AssignmentException(lhs, rhs)
+    if tags is not None and tags != [tag]:
+        evaluation.message(name, "tag", Symbol(name), Symbol(tag))
+        raise AssignmentException(lhs, rhs)
+    if is_protected(tag, evaluation.definitions):
+        evaluation.message(name, "wrsym", Symbol(tag))
+        raise AssignmentException(lhs, None)
+    option_values = rhs.get_option_values(evaluation)
+    if option_values is None:
+        evaluation.message(name, "options", rhs)
+        raise AssignmentException(lhs, None)
+    evaluation.definitions.set_options(tag, option_values)
+    return True
+
+
+def process_assign_attributes(self, lhs, rhs, evaluation, tags, upset):
+    name = lhs.get_head_name()
+    if len(lhs.leaves) != 1:
+        evaluation.message_args(name, len(lhs.leaves), 1)
+        raise AssignmentException(lhs, rhs)
+    tag = lhs.leaves[0].get_name()
+    if not tag:
+        evaluation.message(name, "sym", lhs.leaves[0], 1)
+        raise AssignmentException(lhs, rhs)
+    if tags is not None and tags != [tag]:
+        evaluation.message(name, "tag", Symbol(name), Symbol(tag))
+        raise AssignmentException(lhs, rhs)
+    attributes = get_symbol_list(
+        rhs, lambda item: evaluation.message(name, "sym", item, 1)
+    )
+    if attributes is None:
+        raise AssignmentException(lhs, rhs)
+    if "System`Locked" in evaluation.definitions.get_attributes(tag):
+        evaluation.message(name, "locked", Symbol(tag))
+        raise AssignmentException(lhs, rhs)
+    evaluation.definitions.set_attributes(tag, attributes)
+    return True
+
+
+def process_assign_n(self, lhs, rhs, evaluation, tags, upset):
+    lhs, condition = unroll_conditions(lhs)
+    lhs, rhs = unroll_patterns(lhs, rhs, evaluation)
+    defs = evaluation.definitions
+
+    if len(lhs.leaves) not in (1, 2):
+        evaluation.message_args("N", len(lhs.leaves), 1, 2)
+        raise AssignmentException(lhs, None)
+    if len(lhs.leaves) == 1:
+        nprec = SymbolMachinePrecision
+    else:
+        nprec = lhs.leaves[1]
+    focus = lhs.leaves[0]
+    lhs = Expression(SymbolN, focus, nprec)
```

Comment:
Maybe `apply_N` could be used here?

Reply:
Nop: In this case, we need to keep the `N` unevaluated, because `apply_elementary` could be called both from `Set` as from `DelayedSet`.